pkg/Library/NodeGraph/Editor.js

```diff
@@ -1,193 +1,614 @@
+ * license that can be found in the LICENSE file.
  */
- ({
-  catalogDelimiter: '$$',
+({
+catalogDelimeter: '$$',
+edgeIdDelimeter: '$$',
 
-  update({pipeline, selectedNode}, state) {
-    if (pipeline?.$meta?.name !== state.selectedPipelineName) {
-      state.selectedPipelineName = pipeline?.$meta.name;
-      selectedNode = null;
+update({pipeline, selectedNode, nodeTypes}, state) {
+  state.nodeTypeMap = mapBy(nodeTypes, n => n?.$meta?.name);
+  if (pipeline?.$meta?.name !== state.selectedPipelineName) {
+    state.selectedPipelineName = pipeline?.$meta.name;
+    // new pipeline, choose a selectedNode if there isn't one
+    selectedNode = selectedNode || pipeline?.nodes?.[0];
+  }
+  selectedNode = this.processPostDuplicationData({selectedNode, pipeline}, state);
+  return {
+    selectedNode,
+    ...this.processPostAddCandidateData({selectedNode, pipeline}, state)
+  };
+},
+
+/**
+ * Post-processes the nodes that were just duplicated from the node graph
+ * editor.
+ */
+processPostDuplicationData({selectedNode, pipeline}, state) {
+  const {postDuplicationData: data} = state;
+  // Set the newly duplicated nodes as the local selected nodes in graph.
+  // Will be `undefined` if `!postDuplicationData`
+  state.graphLocalSelectedNodeKeys = data?.graphLocalSelectedNodeKeys;
+  // do nothing else without data
+  if (data) {
+    // Set the main selected node.
+    selectedNode = pipeline.nodes.find(
+      ({key}) => key === data.newSelectedNodeKey
+    );
+    state.postDuplicationData = null;
+  }
+  return selectedNode;
+},
+
+/**
+ * Post-processes the data for the candidate that was just added when user drops
+ * an edge from a node's output onto the empty space in node graph editor.
+ */
+processPostAddCandidateData({pipeline}, state) {
+  if (state.postAddCandidateData) {
+    const {newNodeKey, fromKey, fromStore, toStoreType, nodeType} =
+        state.postAddCandidateData;
+    const node = this.findNodeByKey(newNodeKey, pipeline);
+    // Wait for the connections data to be populated.
+    if (!node.connections) {
+      return;
     }
-    if (!selectedNode && pipeline?.nodes?.length) {
-      selectedNode = pipeline.nodes[0];
+    // Connect to the first store that matches toStoreType.
+    // TODO(b/244191110): Type matching API to be wired here.
+    const toStore = keys(nodeType.$stores).find(
+      storeName => nodeType.$stores[storeName].$type === toStoreType
+    );
+    const toConnection = node.connections[toStore].candidates.find(
+      candidate => candidate.from === fromKey
+    );
+    if (!toConnection?.selected) {
+      pipeline = this.updateStoreConn(
+        pipeline,
+        {fromKey, fromStore, toKey: newNodeKey, toStore},
+        true
+      );
+      return {pipeline};
+    } else {
+      state.postAddCandidateData = undefined;
     }
-    return {selectedNode};
-  },
+    // TODO(jingjin): put the runner UI for the just-added candidate node next
```

Comment:
optional - do we want to rename the todos?

Reply:
Unclear, let's discuss.